src/hypothesis/extra/datetime.py

```diff
@@ -15,111 +15,412 @@
+            max_dt = self.max_dt
+        else:
+            # Choose a timezone to work in on this draw
+            if cu.boolean(data):
+                min_dt = self.min_dt
+                try:
+                    max_dt = self.max_dt.astimezone(self.min_dt.tzinfo)
+                except OverflowError:  # pragma: no branch  # pytest internals
+                    max_dt = self.min_dt.tzinfo.localize(dt.datetime.max)
+            else:
+                max_dt = self.max_dt
+                try:
+                    min_dt = self.min_dt.astimezone(self.max_dt.tzinfo)
+                except OverflowError:
+                    min_dt = self.max_dt.tzinfo.localize(dt.datetime.min)
+
+        result = self._draw_naive(data, min_dt, max_dt)
+        if (not self.timezones) or (self.allow_naive and cu.boolean(data)):
+            assert self.min_dt.tzinfo is None
+            return result
+        timezones = list(self.timezones)
+        for _ in range(len(self.timezones)):
+            try:
+                to_tz = cu.choice(data, timezones)
+                if self.min_dt.tzinfo is None:
+                    return to_tz.normalize(to_tz.localize(result))
+                else:
+                    return to_tz.normalize(result.astimezone(to_tz))
+            except OverflowError:
+                timezones.remove(to_tz)
+        raise InvalidState('Reached impossible state')  # pragma: no cover
 
 
 @defines_strategy
-def datetimes(allow_naive=None, timezones=None, min_year=None, max_year=None):
+def datetimes(allow_naive=None, timezones=None, min_year=None, max_year=None,
+              min_datetime=None, max_datetime=None):
     """Return a strategy for generating datetimes.
 
     allow_naive=True will cause the values to sometimes be naive.
     timezones is the set of permissible timezones. If set to an empty
-    collection all timezones must be naive. If set to None all available
+    collection all timezones will be naive. If set to None all available
     timezones will be used.
 
+    There are two ways to set bounds on the values from this strategy:
+    timezone-aware and naive.
+
+    If two timezone-aware datetime objects are used as the bounds, all
+    values drawn from the strategy will be timezone-aware.  The moment
+    in time represented by a drawn value will be between the moments
+    represented by the bounds.  Because values may be represented in
+    any of the supplied timezones, a naive comparison may show that a value
+    is out of bounds - for example, 10am in London is actually earlier than
+    8am in New York.
+
+    If two naive date or datetime objects are used as the bounds, values
+    drawn from the strategy may or may not be timezone-aware.
+    All values will lie between the bounds when their tzinfo is ignored.
+    If a date object is used, time information will be added to allow the
+    widest possible range of values.
+
+    The min_year and max_year arguments are deprecated, but if given they
```

Comment:
I think Sphinx has a `.. deprecated::` directive that lets you include a version.